src/main/ruby/core/io.rb

```diff
@@ -140,7 +141,12 @@ def exhausted?
     # the IO, and +unshift+ again beginning at +start_pos+.
     def unshift(str, start_pos)
       @write_synced = false
-      written = Truffle.invoke_primitive :iobuffer_unshift, self, str, start_pos, @used
+      available = SIZE - @used
+      written = str.bytesize - start_pos
+      if written > available
+        written = available
+      end
+      @storage.fill(@used, str, start_pos, written)
```

Comment:
I've read this a few times over and I can't understand why `iobuffer_fill` sets the length of `@storage`, but `unshift` does not. `unshift` appears to ignore the `@storage` length and just uses its upper-bound preallocated size of 32,768, then updates the `@used` value, which may now be larger than `@storage.length`.

Reply:
I don't know either, should be cleaned up when the rest of InternalBuffer is cleaned up.